src/protocol.rs

```diff
@@ -1,34 +1,156 @@
+}
+
+// Used for params that are smaller than 255 bytes
+pub struct NinaSmallArrayParam {
+    length: u8,
+    data: Vec<u8, MAX_NINA_PARAM_LENGTH>,
+}
+// Used for params that can be larger than 255 bytes up to MAX_NINA_PARAM_LENGTH
+pub struct NinaLargeArrayParam {
+    length: u16,
+    data: Vec<u8, MAX_NINA_PARAM_LENGTH>,
+}
+
+impl NinaParam for NinaByteParam {
+    type LengthAsBytes = [u8; 1];
+
+    fn new(data: &str) -> Self {
+        let data_as_bytes: Vec<u8, 1> = String::from(data).into_bytes();
+        Self {
+            length: data_as_bytes.len() as u8,
+            data: data_as_bytes,
+        }
+    }
+
+    fn data(&mut self) -> &[u8] {
+        self.data.as_slice()
+    }
+
+    fn length_as_bytes(&mut self) -> Self::LengthAsBytes {
+        [self.length as u8]
+    }
+}
+
+impl NinaParam for NinaWordParam {
+    type LengthAsBytes = [u8; 1];
+
+    fn new(data: &str) -> Self {
+        let data_as_bytes: Vec<u8, 2> = String::from(data).into_bytes();
+        Self {
+            length: data_as_bytes.len() as u8,
+            data: data_as_bytes,
+        }
+    }
+
+    fn data(&mut self) -> &[u8] {
+        self.data.as_slice()
+    }
+
+    fn length_as_bytes(&mut self) -> Self::LengthAsBytes {
+        [self.length as u8]
+    }
+}
+
+impl NinaParam for NinaSmallArrayParam {
+    type LengthAsBytes = [u8; 1];
+
+    fn new(data: &str) -> Self {
+        let data_as_bytes: Vec<u8, MAX_NINA_PARAM_LENGTH> = String::from(data).into_bytes();
+        Self {
+            length: data_as_bytes.len() as u8,
+            data: data_as_bytes,
+        }
+    }
+
+    fn data(&mut self) -> &[u8] {
+        self.data.as_slice()
+    }
+
+    fn length_as_bytes(&mut self) -> Self::LengthAsBytes {
+        [self.length as u8]
+    }
+}
+
+impl NinaParam for NinaLargeArrayParam {
+    type LengthAsBytes = [u8; 2];
+
+    fn new(data: &str) -> Self {
+        let data_as_bytes: Vec<u8, MAX_NINA_PARAM_LENGTH> = String::from(data).into_bytes();
+        Self {
+            length: data_as_bytes.len() as u16,
+            data: data_as_bytes,
+        }
+    }
+
+    fn data(&mut self) -> &[u8] {
+        self.data.as_slice()
+    }
+
+    fn length_as_bytes(&mut self) -> Self::LengthAsBytes {
+        [
+            ((self.length & 0xff00) >> 8) as u8,
+            (self.length & 0xff) as u8,
+        ]
+    }
 }
 
 pub trait ProtocolInterface {
     fn init(&mut self);
     fn reset<D: DelayUs>(&mut self, delay: &mut D);
     fn get_fw_version(&mut self) -> Result<FirmwareVersion, self::Error>;
+    fn set_passphrase(&mut self, ssid: &str, passphrase: &str) -> Result<(), Error>;
+    fn get_conn_status(&mut self) -> Result<u8, self::Error>;
 
     fn send_cmd(&mut self, cmd: NinaCommand, num_params: u8) -> Result<(), self::Error>;
     fn wait_response_cmd(
         &mut self,
         cmd: NinaCommand,
         num_params: u8,
-    ) -> Result<[u8; PARAMS_ARRAY_LEN], self::Error>;
+    ) -> Result<[u8; 8], self::Error>;
```

Comment:
Do you have a specific reason why this switched back to a literal 8 size instead of a constant?

Reply:
Good question. Originally we had the constant `PARAMS_ARRAY_LEN` which is now `MAX_NINA_PARAM_LENGTH` and set to 255. 

255 seemed a little much for this part of the system to I switched it to 8 with the understanding that we'll likely not use `wait_response_cmd` for much longer. I figured we'd come up with a more reasonable way to type this when improving things.

Reply:
Yeah that makes sense. How about creating a new temporary constant that you use everywhere you have an 8 right now? ...at least then it'll be an obvious placeholder for all instances that need to change.

Reply:
Yup! love that idea. f1f275e24eca2a96f66e6cee52a98c6a99aa69d2